p4include/psa.p4

```diff
@@ -25,403 +24,518 @@ limitations under the License.
+  void emit<T>(in T hdr);
+}
+// END:Clone_extern
 
-/// Supported hash algorithms
-enum HashAlgorithm {
-  crc32,
-  crc32_custom,
-  crc16,
-  crc16_custom,
-  random,
-  identity
-}
-
-/// Hash function
-///
-/// @param Algo The algorithm to use for computation (@see HashAlgorithm).
-/// @param O    The type of the return value of the hash.
-///
-/// Example usage:
-/// parser P() {
-///   Hash<crc16, bit<56>> h();
-///   bit<56> hash_value = h.getHash(16, buffer, 100);
-///   ...
-/// }
-extern Hash<Algo, O> {
-  /// Constructor
-  Hash();
 
-  /// compute the hash for data
-  O getHash<T, D, M>(in T base, in D data, in M max);
+// BEGIN:Resubmit_extern
+extern resubmit {
+  /// Write @hdr into the ingress packet buffer.
+  /// @T can be a header type, a header stack, a header union or a struct
+  /// containing fields with such types.
+  void emit<T>(in T hdr);
+}
+// END:Resubmit_extern
+
+// BEGIN:Recirculate_extern
+extern recirculate {
+  /// Write @hdr into the egress packet.
+  /// @T can be a header type, a header stack, a header union or a struct
+  /// containing fields with such types.
+  void emit<T>(in T hdr);
+}
+// END:Recirculate_extern
+
+// BEGIN:Hash_algorithms
+enum HashAlgorithm_t {
+  IDENTITY,
+  CRC32,
+  CRC32_CUSTOM,
+  CRC16,
+  CRC16_CUSTOM,
+  ONES_COMPLEMENT16,  /// One's complement 16-bit sum used for IPv4 headers,
+                      /// TCP, and UDP.
+  TARGET_DEFAULT      /// target implementation defined
+}
+// END:Hash_algorithms
 
-  /*
-  @ControlPlaneAPI
-  { }
-  */
+// BEGIN:Hash_extern
+extern Hash<O> {
+  /// Constructor
+  Hash(HashAlgorithm_t algo);
+
+  /// Compute the hash for data.
+  /// @param data The data over which to calculate the hash.
+  /// @return The hash value.
+  O get_hash<D>(in D data);
+
+  /// Compute the hash for data, with modulo by max, then add base.
+  /// @param base Minimum return value.
+  /// @param data The data over which to calculate the hash.
+  /// @param max The hash value is divided by max to get modulo.
+  ///        An implementation may limit the largest value supported,
+  ///        e.g. to a value like 32, or 256.
+  /// @return (base + (h % max)) where h is the hash value.
+  O get_hash<T, D>(in T base, in D data, in T max);
 }
+// END:Hash_extern
 
-/* --------------------------------------------------------------------- */
-/// Checksum computation
-///
-/// @param W    The width of the checksum
+// BEGIN:Checksum_extern
 extern Checksum<W> {
-  Checksum(HashAlgorithm hash);          /// constructor
-  void clear();              /// prepare unit for computation
-  void update<T>(in T data); /// add data to checksum
-  void remove<T>(in T data); /// remove data from existing checksum
-  W    get();      	     /// get the checksum for data added since last clear
+  /// Constructor
+  Checksum(HashAlgorithm_t hash);
 
-  /*
-  @ControlPlaneAPI
-  { }
-  */
+  /// Reset internal state and prepare unit for computation
+  void clear();
```

Comment:
Is this necessary if the architecture does it?

Reply:
I will create an issue to resolve this question.  Right now InternetChecksum has a clear method, but the only reason you are ever required to call it in a correct program is if you want to reuse the same instance for calculating two different checksums for the same packet.  I suspect Checksum's clear() method can be documented similarly.